plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/HostedHandlerSupport.java

```diff
@@ -0,0 +1,106 @@
+ * of Sonatype, Inc. Apache Maven is a trademark of the Apache Software Foundation. M2eclipse is a trademark of the
+ * Eclipse Foundation. All other trademarks are the property of their respective owners.
+ */
+package org.sonatype.nexus.repository.maven.internal;
+
+import java.io.IOException;
+import java.util.Collections;
+
+import javax.annotation.Nonnull;
+
+import org.sonatype.nexus.common.hash.HashAlgorithm;
+import org.sonatype.nexus.common.time.Clock;
+import org.sonatype.nexus.repository.content.InvalidContentException;
+import org.sonatype.nexus.repository.http.HttpResponses;
+import org.sonatype.nexus.repository.view.Context;
+import org.sonatype.nexus.repository.view.Handler;
+import org.sonatype.nexus.repository.view.Response;
+import org.sonatype.sisu.goodies.common.ComponentSupport;
+
+import com.google.common.hash.HashCode;
+import org.joda.time.DateTime;
+
+import static org.sonatype.nexus.repository.http.HttpMethods.DELETE;
+import static org.sonatype.nexus.repository.http.HttpMethods.GET;
+import static org.sonatype.nexus.repository.http.HttpMethods.PUT;
+import static org.sonatype.nexus.repository.maven.internal.MavenPayload.ok;
+
+/**
+ * Maven hosted handler base.
+ *
+ * @since 3.0
+ */
+public abstract class HostedHandlerSupport<C extends Path>
+    extends ComponentSupport
+    implements Handler
+{
+  private final Clock clock = new Clock();
+
+  @Nonnull
+  @Override
+  public Response handle(final @Nonnull Context context) throws Exception {
+    final C path;
+    try {
+      path = getPath(context);
+    }
+    catch (IllegalArgumentException e) {
+      return HttpResponses.badRequest(e.getMessage());
+    }
+    final String action = context.getRequest().getAction();
+    switch (action) {
+      case GET: {
+        final MavenPayload content = get(context, path);
+        if (content == null) {
+          return HttpResponses.notFound(path.getPath());
+        }
+        return ok(content);
+      }
+
+      case PUT: {
+        try {
+          final DateTime now = clock.dateTime();
+          final MavenPayload mavenPayload = new MavenPayload(
+              context.getRequest().getPayload(),
+              now,
+              now,
+              Collections.<HashAlgorithm, HashCode>emptyMap()
+          );
+          put(context, path, mavenPayload);
+          return HttpResponses.created();
+        }
+        catch (InvalidContentException e) {
+          return HttpResponses.badRequest(e.getMessage());
+        }
+      }
+
+      case DELETE: {
+        final boolean deleted = delete(context, path);
+        if (!deleted) {
+          return HttpResponses.notFound(path.getPath());
+        }
+        return HttpResponses.noContent();
+      }
+
+      default:
+        return HttpResponses.methodNotAllowed(context.getRequest().getAction(), GET, PUT, DELETE);
+    }
+  }
+
+  protected abstract C getPath(final Context context) throws IllegalArgumentException;
```

Comment:
You should not _need_ to handle a RuntimeException that is part of the entire concept related to that classification of exception.  If you want a reminder add javadoc.